### PR TITLE
[Bugfix] [Core] Fix zero temperature case (#5404 and part of #5898)

### LINKED ISSUE #5404: [Bug]:  topk=1 and temperature=0 cause different output in vllm
### 🐛 Describe the bug

When using different generation configurations, such as top_k=1 or temperature=0 (while keeping other settings unchanged), why do the generated results change? They should both correspond to a deterministic greedy decoding.
vllm 0.4.3

---
Supplement:

The main issue encountered here is that the results generated by setting the temperature coefficient to 0 or topk to 1 are different. I understand that due to operator optimization and the lack of conventional arithmetic properties in floating-point numbers, matrix operations have a certain randomness. However, the sampling process occurs after the hidden_state is generated, at which point no calculations are involved. Therefore, the sampling results of the two sampling parameters should be the same.

### LINKED ISSUE #5898: [Bug]: Inconsistent Responses with VLLM When Batch Size > 1 even temperature = 0
### Your current environment

```text
The output of `python collect_env.py`
```

### 🐛 Describe the bug

Test Environment:
- Hardware: A100 80GB GPU
- Model: Llama3-8b 
- Parameters: temperature = 0, max_tokens = 1024, max_num_seqs = 256, seed=1
- I make OpenAI-Compatilbe Server using python -m vllm.entrypoints.openai.api_server  ...

Test Method:
- First, requests were sent one at a time to verify if the same prompt consistently produces the same response.
- Second, more than one requests with same prompt at a time were sent  to verify if the same prompt consistently produces the same response.

Discovered Issue:
- Consistency with Single Request: When the batch size is 1, the same prompt consistently produces the same response.
- Inconsistency with Multiple Requests: When the batch size increases to more than 1, the responses for the same prompt become inconsistent.
- When I set vllm server with parameter max_num_seqs = 1, the result is all same

I think this suggests that the issue arises from the way the Batch Scheduler processes multiple requests together.

My problem is similar with others
https://github.com/vllm-project/vllm/issues/3544
https://github.com/vllm-project/vllm/issues/4606
https://github.com/vllm-project/vllm/issues/608

### PR DESCRIPTION
**Pull request by StanHatko**
Fixes nondeterminism from the sampler that occurs even at zero temperature. The patch supports both zero and nonzero temperatures for different entries in the same batch, using the correct operation in each case. It does this by multiplying by an is-zero indicator for the zero temperature calculation and is-nonzero indicator for positive temperature case. These two are then summed. For any entry, only one term in the sum is nonzero as the two cases are mutually exclusive.

I created this patch and discussed it in https://github.com/vllm-project/vllm/issues/5404#issuecomment-2628860618. I used this patch internally and it greatly reduced variation in answers at zero temperature (remaining nondeterminism I think is due to CUDA nondeterminacy in the steps leading up to calculation of the logits).

Please feel free to make any improvements to this patch you can think of.

FIX #5404 (patch discussed there).

FIX #5898 (part from sampler, nondeterminism from the computation of the logits themselves not covered here).

### CODE REVIEW

**Review by ywang96 — COMMENTED** [review]
Very much appreciated the investigation! IMO this is quite an elegant patch - @WoosukKwon What do you think?

**Comment by StanHatko** [comment]
For the pre-commit hook error here, it shows missing SPDX headers in the file `tests/lora/test_ultravox.py` which I didn't modify at all in my commits. The pre-commit hooks passed on my computer, so I think that was unrelated commits by others to vLLM since.

For the test with increased GPU VRAM usage, some increase makes sense due to the creation of additional tensors. The previous code just did an in-place divide after the conversion to float32, while the current code needs to compute tensors for both cases and sum them. I think correctness of outputs at zero temperature (a very common use case) is more important though than saving GPU VRAM at the sampling step. Also it may be possible to optimize the code to do more stuff in-place, but some new tensors are inevitable I think with doing it this way.

However, it's possible to avoid the increased GPU RAM usage with a custom CUDA kernel. Basically re-write this as a CUDA kernel and call that. No new tensors will be needed as the computations will be done in temporary variables within a loop, with the final output values assigned directly to the output tensor. I think we should first merge this though and then implement the CUDA kernel as an optimization. I have no experience writing CUDA kernels (though I did write and optimize OpenCL code many years ago).

**Review by Muhtasham — APPROVED** [review]

**Comment by WoosukKwon** [comment]
Wow this is amazing! Thanks for the thorough investigation!

**Comment by WoosukKwon** [comment]
Sorry for the delay in review. LGTM overall, but I'd like to understand the problem in more detail. Let me get back to this within 1~2 days.

**Comment by njhill** [comment]
To me this looks like a workaround to an underlying problem of doing argmax on the logprobs for greedy.

@StanHatko could you try out this PR which fixes that? https://github.com/vllm-project/vllm/pull/13312

Man I forgot how convoluted the v0 sampler code was :sweat_smile:

**Comment by StanHatko** [comment]
Yes fixing the sampler to be deterministic in the temperature = 0 case seems like a better solution, ideally by just taking the largest logit in that case everything here can be avoided. I'll check if that pull request fixes the issue.

**Comment by njhill** [comment]
Thanks @StanHatko 

> If there's a batch with a mixture of zero and nonzero temperatures, will it use greedy sampling for the zero-temperature entries and regular sampling for the rest?

@StanHatko yes that's correct, and greedy requests are determined by `temperature < _SAMPLING_EPS`.
